net, vm ctrl: Adjust VM's detached iface clearing logic

orelmisan opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as a draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Currently, VM's absent interfaces which are not present in the VMI.Status.Interfaces slice, are removed.

This is problematic in the following scenario:

1. The dynamic-networks-controller is not deployed.
2. A VM is running with an interface connected to pod network.
3. A secondary interface is hotplugged and the VM is not migrated.
4. The interface above is marked for hot-unplug by setting its state field to `absent`.

In this case, the interface and its matching network are removed from the VM spec, but are kept in the VMI spec.

This is because, the controller does not set the VMI's interface state as absent, as in this scenario the naming scheme is unknown and assumed to be ordinal, thus preventing hot-unplug.

Adjust the logic for clearing VM interfaces and their matching networks to remove only interfaces that meet the following criteria:
1. The interface state is absent.
2. The interface does not exist in the VMI spec.
3. The interface does not exist in the VMI status.

Following this change, in the scenario described above, the interface and its matching network will be retained both on the VM and VMI specs, until the VM will be migrated.

### References
<!-- optional,
  Use `Fixes #<issue number>(, Fixes #<issue_number>, ...)` format, to close the issue(s) when PR gets merged.
  Use `Partially addresses #<issue number>` to link an issue without closing it when the PR merges.
  
- Fixes #
- Partially addresses #
-->
<!-- optional,
  VEP tracking issue if this PR is implementing one.
  For additional info about VEP tracking issue, see https://github.com/kubevirt/enhancements#process
  
- VEP tracking issue: https://github.com/kubevirt/enhancements/issue/<vep_tracking_issue_number>
-->

### Why we need it and why it was done in this way
The following tradeoffs were made:

The following alternatives were considered:

Links to places where the discussion took place: <!-- optional: slack, other GH issue, mailinglist, ... -->

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least one e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

orelmisan commented:
/cc @nirdothan @oshoval

oshoval commented:
Please consider to split the 4th commit to a follow-up
already reviewed the first 3, and then the 4th was added, the PR already does enough with the first 3

oshoval commented on pkg/network/controllers/vm.go:
```diff
@@ -106,6 +94,29 @@ func (v *VMController) Sync(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstanc
 	return vmCopy, nil
 }
 
+func (v *VMController) syncVMI(
```
`syncVMIInterfaces` ?

orelmisan replied:
Done.

orelmisan commented on pkg/network/controllers/vm.go:
```diff
@@ -72,21 +72,30 @@ func (v *VMController) Sync(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstanc
 
 	vmCopy := vm.DeepCopy()
-	ifaces, networks := clearDetachedInterfaces(
+	ifaces, networks := clearDetachedVMInterfaces(
```
> name is confusing, it does depend on VMI, but at the end it cleans the interfaces / networks from the vm (vmCopy)

Similar logic existed prior to this change, and it was also dependent on info originating from VMI.Status.Interfaces.

> maybe worth to rename to VM and to rename clearDetachedInterfaces to include VMI in the name instead

Will renaming `clearDetachedInterfaces` to `clearDetachedVMIInterfaces` work?

orelmisan commented on pkg/network/controllers/vm.go:
```diff
@@ -92,14 +92,14 @@ func (v *VMController) Sync(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstanc
 	}
 
 	vmiCopy := vmi.DeepCopy()
-	ifaces, networks = clearDetachedInterfaces(vmiCopy.Spec.Domain.Devices.Interfaces, vmiCopy.Spec.Networks, indexedStatusIfaces)
```
Currently, the following flow happens:
1. Interfaces with an `absent` state and no status are removed from the VM and VMI.
2. Absent interfaces are synced from VM to VMI.
3. VMI is patched.

It is possible that additional interfaces could have been removed after step 2.
These interfaces will only be handled by the next reconcile.

orelmisan commented:
[Change](https://github.com/kubevirt/kubevirt/compare/0a50b9b2106f4a363a7435fc436b059473020339..27d7c64f23f1a61404f46321d2b72b01723d1e89): Addressed review comments.

oshoval commented:
Thanks

/lgtm

oshoval commented on pkg/network/controllers/vm_test.go:
```diff
@@ -608,6 +608,56 @@ var _ = Describe("VM Network Controller", func() {
+		updatedVM, err := c.Sync(vm, vmi)
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(updatedVM.Spec.Template.Spec.Networks).To(Equal(originalVM.Spec.Template.Spec.Networks))
```
i am bit confused, the interface is marked absent
so it should be removed with its network because it was unplugged at least according title
(unless it is retained by https://github.com/kubevirt/kubevirt/pull/14914/files#diff-ce73e9866c7d67ddd64b01df072b2914cb040e8cd5795c7abe4803378bb17097R211 because the test need to not match this condition)

orelmisan replied:
From the last commit message:
> This is because, the controller does not set the VMI's interface state
as absent, as in this scenario the naming scheme is unknown and assumed
to be ordinal, thus preventing hot-unplug.

EdDev commented on pkg/network/controllers/vm.go:
```diff
@@ -106,6 +94,29 @@ func (v *VMController) Sync(vm *v1.VirtualMachine, vmi *v1.VirtualMachineInstanc
+	vmiCopy.Spec.Domain.Devices.Interfaces = ifaces
+	vmiCopy.Spec.Networks = networks
+
+	if err := v.vmiInterfacesPatch(&vmiCopy.Spec, vmi); err != nil {
```
Pushing down the stack the patching operation is making things worse in regards to "hiding" it.
Patching the VMI is a bug and expensive deal, we should try and expose it on the main flow, not push it down the stack and hide it in the low details.

Nevertheless, we could try and handle it as a follow up.

EdDev commented on pkg/network/controllers/vm_test.go:
```diff
@@ -608,6 +608,56 @@ var _ = Describe("VM Network Controller", func() {
 		Expect(updatedVMI.Spec.Domain.Devices.Interfaces).To(Equal(originalVM.Spec.Template.Spec.Domain.Devices.Interfaces))
 	})
+
+	It("Could unplug an interface that is pending hotplug", func() {
```
The name does not describe in full the scenario and it drifts from the used pattern in all other tests here.
It would be nicer to keep the format for consistency and add a bit more information.

e.g. "sync does not hotunplug interfaces when a pending hotplug without a status entry exists"

EdDev reviewed (commented):
Thanks.

I had some minor comments which you could handle as a follow up if you prefer.
I leave it to you to decide, feel free to unhold.

/approve
/hold

orelmisan commented:
/hold cancel

orelmisan commented:
Failures seem unrelated
/retest-required
